cirq/circuits/frozen_circuit.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An immutable version of the Circuit data structure."""
+
+from typing import TYPE_CHECKING, Callable, Sequence, Tuple
+
+from cirq import devices, ops
+from cirq.circuits import AbstractCircuit, Circuit
+from cirq.circuits.insert_strategy import InsertStrategy
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class FrozenCircuit(AbstractCircuit):
+    """An immutable version of the Circuit data structure.
+
+    FrozenCircuits are immutable (and therefore hashable), but otherwise behave
+    identically to regular Circuits. Conversion between the two is handled with
+    the `freeze` and `unfreeze` methods from AbstractCircuit.
+    """
+
+    def __init__(self,
+                 *contents: 'cirq.OP_TREE',
+                 strategy: 'cirq.InsertStrategy' = InsertStrategy.EARLIEST,
+                 device: 'cirq.Device' = devices.UNCONSTRAINED_DEVICE) -> None:
+        base = Circuit(contents, strategy=strategy, device=device)
+        self._moments = tuple(base.moments)
+        self._device = base.device
+
+        # These variables are memoized when first requested.
+        self._num_qubits = None
+        self._qid_shape = None
+        self._all_qubits = None
+        self._all_operations = None
+        self._unitary = None
+        self._has_measurements = None
+        self._all_measurement_keys = None
+        self._are_all_measurements_terminal = None
```

Comment:
nit: add type annotations on these members and return type annotations on the corresponding methods.